### PR TITLE
POC - Sorting code into files by functionality

### PR DESCRIPTION
**Pull request by dbarrosop**
@ktbyers @fooelisa @mirceaulinic 

After a short discussion with @mirceaulinic I realized how big the code was getting so I created this PR for the sake of testing a potential solution to sort code into different files by functionality without breaking or adding extra complexity to the current API. Please, take a look and let me know what you think. The general idea is as follows:

Let's say we create a `bgp.py` file where we want to group all the code related to BGP. We would have methods like:

```
def _my_helper_method():
    # do stuff
   return stuff


def get_bgp_neighbors(node):
   # do stuff
   return my_data

def get_bgp_config(node, group='', neighbor=''):
    # do stuff
    return my_data
```

Then, the eos.py code would be:

```
from . import configmgmt
from . import bgp

class EOSDriver(NetworkDriver):
    (...)
    def get_bgp_config(self, group='', neighbor=''):
        return bgp.get_bgp_config(self, group=group, neighbor=neighbor)

    def get_bgp_neighbors(self):
        return bgp.get_bgp_neighbors(self)
```

We maintain the API while getting rid of a flat file with ~1000 SLOC.

Thoughts? A good thing is that this is optional, nothing enforces this so we can sort the code as we need/want/have_time/not_at_all : )

### CODE REVIEW

**Comment by ktbyers** [comment]
@dbarrosop What advantage do you see in doing this? What problem are we trying to solve?

Are you seeing lots of conflicts on merges?

Kirk

**Comment by dbarrosop** [comment]
The main problem I am trying to solve is having a flat file that is hard to
navigate due to all the functions and helper functions. Right now eos.py is
around a 1000 SLOC and growing. I thought that this would be easier to
navigate and to maintain as eos.py just holds the API definition while all
the relevant code is grouped by functionality.

Regarding conflicts, as a matter of fact, yes. I am seeing more and more
conflicts coming but I am not sure if this would address this issue.

**Comment by ktbyers** [comment]
@dbarrosop Yes, I don't really see an issue here that is worth fixing. I haven't really found navigating the ios.py file that difficult (you usually know the method you are looking for or can trace it down pretty quickly).

I think the change would probably make things somewhat harder as with the proposed solution you would have to start searching across multiple files. You also would have to manage imports from various files (and integrate things together in the package **init**.py).

**Comment by mirceaulinic** [comment]
@ktbyers I consider this moment as a good opportunity to discuss about all necessary changes, otherwise we'll end up restructuring over and over again.
In my opinion, all of us are committed to make this library a super powerful tool and it is extremely important to make it scalable as the length of the classes will grow together with the size of the community. My personal expectation is that at least three or perhaps four of of the drivers will be longer than 3000 SLOC in the next few months. Will this be scalable enough?
As discussed with @dbarrosop, I would propose to leave aside all endless discussions regarding the potentially performance added by dividing a huge file into multiple parts and focus for the moment on how we can make the code more readable (again, I do not refer to the current code but to the much more longer code say in one year from now on).
Having more than one single file per driver indeed has its drawbacks too. Let's take the example of PyEZ or Salt that indeed are hard to navigate (and we definitely need to avoid that!) but highly modular and easily extensible.
I do not claim that this is the right way, I do not claim that we have to split it, but doing so I see two big advantages: flexibility and more readable code.
An this is an essential milestone, please let's get involved more into this discussion and come with other ideas/objections/proposals and make sure we'll do the right thing!

**Comment by dbarrosop** [comment]
It's just a way of sorting code. Methods are growing in complexity a we try to do more interesting stuff and they usually come with several private methods. I also expect to start breaking down methods into things a bit more readable, we have some methods with +100 lines of code that are hard to follow, I'd rather break those down into helper methods that I can read and understand what the code is doing. So I guess helpers methods will grow even larger.

My goal here is just to try to make things easier for reviewers and new contributors and I personally think that a flat file with thousands of lines is a bit intimidating.

**Comment by fooelisa** [comment]
I don't really have any strong feelings about that. I agree that the length of the files is starting to make it hard to navigate right now. But also that splitting it up might become hard to navigate in it's own way (I have the same feeling as @mirceaulinic about pyez for example).

If it were me I would probably start by splitting things into eos.py and utils.py, and relocate helper functions as a first step. Then try to shorten the base methods with additional helpers. I don't feel like the split by bgp / config/ etc. is really necessary at this point, but maybe it'll be in the future...

**Comment by dbarrosop** [comment]
Most of the helpers are private functions as they are not supposed to be reused, just improve readability. Check the private methods inside bgp.py, none can be reused by anyone else. And we need more of those private functions to improve readability, half of the methods are nearly impossible to read and understand.

I don't think this is going to hurt navigability. I understand that it's hard to navigate when you have several packages and classes but this is just using different files to sort methods and all the files are in the same folders as the main class and are used very straightforward. Just refer to the example in the description of the PR.

**Comment by ktbyers** [comment]
I looked at this more yesterday and am now more strongly against it. 

The current structure (for each network driver) is to have a single module with a single class and a bunch of methods. Both the eos.py and ios.py driver currently have 28 high-level methods excluding init (junos.py is similar).

This proposal would take the single Python module and class and subdivide the methods across multiple Python files. 

At the Python level we would be converting a single class with methods into a class plus methods plus functions (pulled from several files). The functions would also have shared state (the 'node' object) and we would be passing this node object around between them.

Conceptually we are taking a class with methods (shared state plus actions) and breaking it apart (and starting to pass around shared state again with the 'node' object). In other words, we are breaking apart one of the values that classes/objects provide you. 

Personally, I think this will make the code meaningfully less readable. Also it is much more important to get the data structures correct, than to focus on the number of lines in the file.

On the flip side, I think you might get some value out of reducing merge conflicts from splitting things up into more files. But I don't think it is worth doing it for this.

I am definitely for refactoring the code and improving existing methods to make them be more logical and easier to follow (including creating helper methods).

Here are some other things that we could do that could help subdivide things better:
1. Convert 'staticmethod' to straight functions and place these in external modules. For example, eos_functions.py (or something more general if they can be shared between drivers). 
2. There might be other ways to subdivide the code at the cfgmgmt and getters divide and still retain the class structure.

**Comment by mirceaulinic** [comment]
> Also it is much more important to get the data structures correct, than to focus on the number of lines in the file.
> 
> Arguing that we can simply ignore this (C0302 — max-module-length, recommended be 1000) is equivalent saying that we can as well ignore C0301 (max-line-length) and have super-long lines.

On the other hand @ktbyers come with very interesting arguments and perhaps we should externalise the problem.

> 1. Convert 'staticmethod' to straight functions and place these in external modules. For example, eos_functions.py (or something more general if they can be shared between drivers). 
> 
> In the base maybe?

**Comment by fooelisa** [comment]
The shared ones are in utils, so all that's left in the driver file is not shared anywhere.

I'm :+1: on creating a helper file for each module and increasing the amount of helpers within the driver methods to reduce the class size.

**Comment by ktbyers** [comment]
The key question is whether it makes the code better or worse. If we write worse code because of adhering to PEP8, then it is a bad idea. 

Shortening the lines to 100 chars doesn't require us to change our data structures so pretty big difference there. 

In other words, I think we should focus on what data structures make sense. This is much more fundamental.

**Comment by dbarrosop** [comment]
One of the early decisions I made was that I wanted a simple library. Other network libraries would do the same with 100 different classes: one for the device, another for DeviceCall, another for BGPNeighbor, another for Interface, for VLAN, for SystemEnvironment, etc... and the code would be spread amongst them. Both `pyeos` and `pyez` are great examples for that. And, honestly, I am very happy I did that as the API is super straightforward to use, read, troubleshoot, etc. However, we need a way to better organize code by functionality because I still think it's getting out of hand and it will grow worse. There are at least three problems I want to solve with this:
1. Conflicts. This is a minor issue but reducing the size and the scope of files reduces the chances of conflicts.
2. Fear. I want people to join the project. So far it's been great and I want to keep it like this. If someone finds a bug and realizes it has to patch a file with 2000 SLOC and hundreds of methods he/she is probably going to not do it for fear of breaking something or for pure laziness of having to go through thousands of lines (I have done that myself several times). However, patching a file called bgp.py with 200 SLOC and 10 methods sounds way less scary : )
3. Peer reviewing. The main reason why I wanted to separate the drivers apply to this. Reviewing is getting harder, a lot of functionality being added all the time and half of them require complex regular expressions with a lot of possible branches in the code. It's easier from a reviewer perspective if code is at least sorted by functionality.

> The functions would also have shared state (the 'node' object) and we would be passing this node object around between them.

I partially agree with that statement and I understand what you mean and I see why it's ugly. At least for the configuration pieces as the _state_ of the device changes with every call. However, for getters there is not state changing. Anyway, something we could do to avoid this is:

```
bgp.py
---------
def _my_helper_method():
    # do stuff
   return stuff


def parse_show_run_router_bgp(output):
   # do stuff and use all the helpers necessary to make the code readable
   return my_data


eos.py
--------

from . import bgp

class EOSDriver(NetworkDriver):
    (...)
    def get_bgp_config(self, group='', neighbor=''):
        output = self.device.run_commands('show run router bgp')
        parsed_output = bgp.parse_show_run_router_bgp(output)
        #maybe do more stuff and even call more methods in the bgp module
        return bgp_config
```

In summary, I don't think that grouping functions by functionality makes us write worse code. Code should be easy to navigate and read and right now we are very far from there.

**Comment by mirceaulinic** [comment]
From the perspective of the object-oriented programming, @ktbyers is right in my opinion. I would prefer a clean implementation too (I am not sure if "clean" is quite the right term. I do not mean that the solution proposed would be "dirty"/less clean, but somehow we exploit the flexibility of Python, avoiding hierarchy of classes).
With these said, I think that @dbarrosop's proposal is a very fair compromise.

**Comment by dbarrosop** [comment]
I am closing this PR for the time being. It's clear we will not be implementing this behavior in this exact way as we all agree that from an OOP it's ugly. I will propose another POC where the code is more readable and only the logic of the parsing is moved out of the class.

**Comment by ktbyers** [comment]
@dbarrosop Yes, I like what you proposed above much better as it would still keep the class/method structure intact and just relocate string processing functions to external files.
